File: VS/lab3/Warehouse.cpp
#include "Warehouse.hpp"
#include "SupplyRequestDetail.hpp"
#include "Ammunition.hpp"
#include "Fuel.hpp"
#include <algorithm>


Warehouse::Warehouse(int id, std::string name, WarehouseType type) :
    m_id(id), m_name(name), m_type(type) {
}

int Warehouse::GetId() const {
    return m_id;
}

WarehouseType Warehouse::GetType() const {
    return m_type;
}

const std::string Warehouse::GetName() const {
    return m_name;
}

void Warehouse::SetName(std::string name) {
    m_name = name;
}

void Warehouse::AddResource(std::unique_ptr<Resource> resource) {
    m_resources.push_back(std::move(resource));
}

void Warehouse::AddResources(std::vector<std::unique_ptr<Resource>>& resources) {
    for (auto& resource : resources) {
        AddResource(std::move(resource));
    }
}

void Warehouse::AddEquipment(std::unique_ptr<Equipment> equipment) {
    m_equipments.push_back(std::move(equipment));
}

void Warehouse::AddEquipment(std::vector<std::unique_ptr<Equipment>>& equipment) {
    for (auto& eq : equipment) {
        AddEquipment(std::move(eq));
    }
}

void Warehouse::ProcessSupplyRequestDetails(std::vector<std::unique_ptr<SupplyRequestDetail>>& details) {

    if (details.empty())
        return;

    for (auto& detail : details) {
        switch (detail->GetSupplyType()) {

        case SupplyType::Ammunition:
        case SupplyType::Fuel:
            WriteOff<Resource>(m_resources, detail.get());
            break;

        case SupplyType::Weapon:
        case SupplyType::Vehicle:
            WriteOff<Equipment>(m_equipments, detail.get());
            break;
        }
    }
    RemoveEmptyEntries(m_resources);
    RemoveEmptyEntries(m_equipments);
    RemoveEmptyRequestDetails(details);
}

template<typename T>
void Warehouse::WriteOff(std::vector<std::unique_ptr<T>>& collection, SupplyRequestDetail* detail) {

    float remaining = detail->GetCount();

    for (auto& item : collection) {

        if (item->IsMatches(detail)) {

            remaining -= item->Decrease(remaining);

            if (remaining == 0.0f)
                break;
        }
    }
    detail->SetCount(remaining);
}

template<typename T>
void Warehouse::RemoveEmptyEntries(std::vector<std::unique_ptr<T>>& pointers) {

    auto newEnd = std::remove_if(pointers.begin(), pointers.end(), [](const std::unique_ptr<T>& pointer)
        {
            return pointer == nullptr || pointer->IsEmpty();
        });

    pointers.erase(newEnd, pointers.end());
}

void Warehouse::RemoveEmptyRequestDetails(std::vector<std::unique_ptr<SupplyRequestDetail>>& pointers) {

    auto newEnd = std::remove_if(pointers.begin(), pointers.end(), [](const std::unique_ptr<SupplyRequestDetail>& pointer)
        {
            return pointer == nullptr || pointer->GetCount() == 0.0f;
        });

    pointers.erase(newEnd, pointers.end());
}

File: lab4/src/warehouse/Warehouse.java
package warehouse;

import enums.*;
import java.util.ArrayList;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import request.SupplyRequestDetail;
import warehouse.equipments.Equipment;
import warehouse.resources.Resource;


public class Warehouse {
    int id;
    WarehouseType type;
    String name;
    public ArrayList<Resource> resources = new ArrayList<>();
    public ArrayList<Equipment> equipments = new ArrayList<>();

    public int getId(){
        return id;
    }
    public WarehouseType getType(){
        return type;
    }
    public String getName(){
        return name;
    };
    
    public void setName(String name){
        this.name = name;
    }

    public Warehouse(int id, String name, WarehouseType type){
        this.id = id;
        this.name = name;
        this.type = type;
    }


    public Boolean addResource(Resource resource){
        Boolean hasId = resources.stream().anyMatch(r->r.getId() == resource.id);
        if(!hasId)
            resources.add(resource);
        return !hasId;
    }
    public int addResources(ArrayList<Resource> rs){
        var copy = new ArrayList<Resource>(rs);
        copy.removeIf(i->resources.stream().anyMatch(j->j.getId() == i.getId()));
        var distincted = copy.stream().filter(distinctBy(r->r.getId())).collect(Collectors.toList());
        resources.addAll(distincted);
        return distincted.size();
    }
    public Boolean addEquipment(Equipment equipment){
        Boolean hasId = equipments.stream().anyMatch(e->e.getId() == equipment.getId());
        if(!hasId)
            equipments.add(equipment);
        return !hasId;
    }
    public int addEquipments(ArrayList<Equipment> eq){
        var copy = new ArrayList<Equipment>(eq);
        copy.removeIf(i->equipments.stream().anyMatch(j->j.getId() == i.getId()));
        var distincted = copy.stream().filter(distinctBy(e->e.getId())).collect(Collectors.toList());
        equipments.addAll(distincted);
        return distincted.size();
    }
    public Boolean processSupplyRequestDetails(ArrayList<SupplyRequestDetail> details){
        if (details.size() < 1)
            return false;

        int origSize = details.size();

        for (SupplyRequestDetail detail : details) {
            switch (detail.getSupplyType()) {
                case AMMUNITION, FUEL -> writeOff(resources, detail);
                case WEAPON, VEHICLE->writeOff(equipments, detail);
            }
        }

        removeEmptyEntries(resources);
        removeEmptyEntries(equipments);
        removeEmptyRequestDetails(details);

        return origSize > details.size();
    }

    <T extends WarehouseEntry> Boolean writeOff(ArrayList<T> entries, SupplyRequestDetail detail){

        Boolean isCorrected = false;
        float remaining = detail.getCount();
        for(WarehouseEntry entry: entries){
            if(entry.isMatches(detail)){
                isCorrected = true;
                remaining-=(float)entry.decrease(remaining);
                if (remaining == 0.0f)
                    break;
            }
        }
        detail.setCount(remaining);
        return isCorrected;
    }

    <T extends WarehouseEntry> void removeEmptyEntries(ArrayList<T> entries){
        entries.removeIf(entry-> entry.isEmpty());
    }

    void removeEmptyRequestDetails(ArrayList<SupplyRequestDetail> details) {
        details.removeIf(detail->detail.getCount() == 0.0f);
    }

    static <T> Predicate<T> distinctBy(Function<? super T, ?> keyExtractor) {
        Set<Object> seen = ConcurrentHashMap.newKeySet();
        return t -> seen.add(keyExtractor.apply(t));
    }
}
